core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala

```diff
@@ -66,86 +66,102 @@ private[storage] class BlockManagerDecommissioner(
+          // will finish being committed.
+          case None => Thread.sleep(1000)
+        }
+      }
+      throw new InterruptedException()
+    }
+
     override def run(): Unit = {
-      var migrating: Option[(ShuffleBlockInfo, Int)] = None
-      logInfo(s"Starting migration thread for ${peer}")
+      logInfo(s"Starting shuffle block migration thread for $peer")
       // Once a block fails to transfer to an executor stop trying to transfer more blocks
-      try {
-        while (running && !Thread.interrupted()) {
-          migrating = Option(shufflesToMigrate.poll())
-          migrating match {
-            case None =>
-              logDebug("Nothing to migrate")
-              // Nothing to do right now, but maybe a transfer will fail or a new block
-              // will finish being committed.
-              val SLEEP_TIME_SECS = 1
-              Thread.sleep(SLEEP_TIME_SECS * 1000L)
-            case Some((shuffleBlockInfo, retryCount)) =>
-              if (retryCount < maxReplicationFailuresForDecommission) {
-                val blocks = bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo)
-                if (blocks.isEmpty) {
-                  logInfo(s"Ignore empty shuffle block $shuffleBlockInfo")
+      while (keepRunning) {
+        try {
+          val (shuffleBlockInfo, retryCount) = fetchNextShuffleBlock()
+          val blocks = bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo)
+          if (blocks.isEmpty) {
+            logInfo(s"Ignore empty shuffle block $shuffleBlockInfo")
+          } else {
+            logInfo(s"Got migration sub-blocks $blocks. Trying to migrate $shuffleBlockInfo " +
+              s"to $peer ($retryCount / $maxReplicationFailuresForDecommission)")
+            // Migrate the components of the blocks.
+            try {
+              blocks.foreach { case (blockId, buffer) =>
+                logDebug(s"Migrating sub-block ${blockId}")
+                bm.blockTransferService.uploadBlockSync(
+                  peer.host,
+                  peer.port,
+                  peer.executorId,
+                  blockId,
+                  buffer,
+                  StorageLevel.DISK_ONLY,
+                  null) // class tag, we don't need for shuffle
+                logDebug(s"Migrated sub block $blockId")
+              }
+              logInfo(s"Migrated $shuffleBlockInfo to $peer")
+            } catch {
+              case e: IOException =>
+                // If a block got deleted before netty opened the file handle, then trying to
+                // load the blocks now will fail. This is most likely to occur if we start
+                // migrating blocks and then the shuffle TTL cleaner kicks in. However this
+                // could also happen with manually managed shuffles or a GC event on the
+                // driver a no longer referenced RDD with shuffle files.
+                if (bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).isEmpty) {
```

Comment:
I am just wondering whether it makes sense to change this condition to 

```
bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).size < blocks.size
```

As `getMigrationBlocks` currently gives back the shuffle data and index block ID and if the data is already lost (index cannot be lost on its own) we already know a deletion is in progress. WDYT?

Reply:
Good point! Changed to `bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).size < 2`. (This makes the condition fit in one single line and I believe it's the same as you suggested here)

And I also changed the condition from `blocks.isEmpty` to `blocks.size < 2` for the same reason:
https://github.com/apache/spark/blob/8c5fecda7320dce32fc36981c41407f4272a7748/core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala#L87

Reply:
As it's possible to have index-only shuffle files (https://github.com/apache/spark/pull/32727), so I have reverted the changes made here.